### PR TITLE
Fixed entity linking for training on wikitables

### PR DESCRIPTION
**Pull request by pdasigi**
This PR is missing two things:
1) I still need to fix merge conflicts, most of which are due to the recent changes in DyNet.
2) I need to add numbers as entities. This doesn't seem very straightforward. I may need to change the type declaration a bit for this.

### CODE REVIEW

**Comment by jayantk** [comment]
Can you merge the master branch with this one? I think that will fix most of you dynet problems. I'm developing most PNP stuff on master to avoid the dependency on sempre.

File: src/main/scala/org/allenai/pnp/semparse/EntityLinking.scala [hunk 1]
```diff
@@ -7,10 +7,14 @@ import com.jayantkrish.jklol.ccg.lambda2.Expression2
 import scala.collection.mutable.ListBuffer
 
 case class EntityLinking(matches: List[(Span, Entity, List[Int], Double)]) {
-  val entities = matches.map(_._2).toSet.toList
-  
+  val entities: List[Entity] = matches.map(_._2)
+
+  // Some matches may have null span. They correspond to floating rules. The following are the rest.
+  val linkedMatches = matches.filter(x => x._1 != null)
```
Comment (jayantk) [inline src/main/scala/org/allenai/pnp/semparse/EntityLinking.scala]: Using null isn't very scala like. Use Option[Span] instead

File: src/main/scala/org/allenai/pnp/semparse/SemanticParser.scala [hunk 2]
```diff
@@ -126,10 +127,14 @@ class SemanticParser(val actionSpace: ActionSpace, val vocab: IndexedList[String
...
       }
       Preconditions.checkState(lastOutput != null)
-      
+      // Entity encoding is average of the representations of the corresponding span.
+      lastOutput = lastOutput / name.length
       val spanVector = new FloatVector(tokens.length)
       for (i <- 0 until tokens.length) {
-        val value = if (i >= span.start && i < span.end) { 1.0 } else { 0.0 }
+        val spanStart = if(span == null) -1 else span.start
+        val spanEnd = if(span == null)  -1 else span.end
```
Comment (jayantk) [inline src/main/scala/org/allenai/pnp/semparse/SemanticParser.scala]: this will change too if you use Option[Span]

Comment (jayantk) [inline src/main/scala/org/allenai/pnp/semparse/SemanticParser.scala]: Also put the check for whether the span exists outside the loop. The logic here is confusing.

File: src/main/scala/org/allenai/pnp/semparse/SemanticParser.scala [hunk 3]
```diff
@@ -268,13 +273,12 @@ class SemanticParser(val actionSpace: ActionSpace, val vocab: IndexedList[String
             entityBias <- param(SemanticParser.ENTITY_BIAS_PARAM + hole.t)
             entityWeights <- param(SemanticParser.ENTITY_WEIGHTS_PARAM + hole.t)
             entityChoiceScore = dot_product(entityWeights, rnnOutputDropout) + entityBias 
-            // TODO: How should we score these entities using attentions?
-            /*
+            // TODO: How should we score these entities using attentions
```
Comment (jayantk) [inline src/main/scala/org/allenai/pnp/semparse/SemanticParser.scala]: I believe this version of entity scoring worked worse than the one below on GeoQuery. Something to keep in mind when we run experiments.

File: src/main/scala/org/allenai/wikitables/WikiTablesSemanticParserCli.scala [hunk 4]
```diff
@@ -49,38 +49,43 @@ class WikiTablesSemanticParserCli extends AbstractCli() {
   
   var trainingDataOpt: OptionSpec[String] = null
   var testDataOpt: OptionSpec[String] = null
-  
+  val typeDeclaration = new WikiTablesTypeDeclaration()
```
Comment (jayantk) [inline src/main/scala/org/allenai/wikitables/WikiTablesSemanticParserCli.scala]: Declare this inside run() and pass it to the preprocessExamples method. That way we can swap in different type declarations if necessary.

File: src/main/scala/org/allenai/wikitables/WikiTablesSemanticParserCli.scala [hunk 5]
```diff
@@ -170,19 +190,31 @@ class WikiTablesSemanticParserCli extends AbstractCli() {
...
     val builder = ListBuffer[(Span, Entity, List[Int], Double)]()
     for (linking <- sempreEntityLinking.asScala) {
-      val entityExpr = Expression2.constant(linking.getSecond().toString)
-      val entityType = WikiTablesTypeDeclaration.CELL_TYPE
+      val entityString = linking.getSecond.toString
+      val entityExpr = Expression2.constant(entityString)
+      val entityType = typeDeclaration.getType(entityString)
       val template = ConstantTemplate(entityType, entityExpr)
       // TODO: Passing names as null. Not sure what happens.
       val entity = Entity(entityExpr, entityType, template, null)
```
Comment (jayantk) [inline src/main/scala/org/allenai/wikitables/WikiTablesSemanticParserCli.scala]: The code below calculates at least one name for entity. Why not fill in the value?

File: src/main/scala/org/allenai/wikitables/WikiTablesSemanticParserCli.scala [hunk 6]
```diff
@@ -170,19 +190,31 @@ class WikiTablesSemanticParserCli extends AbstractCli() {
     val sempreEntityLinking = WikiTablesDataProcessor.getEntityLinking(ex)
     val builder = ListBuffer[(Span, Entity, List[Int], Double)]()
     for (linking <- sempreEntityLinking.asScala) {
-      val entityExpr = Expression2.constant(linking.getSecond().toString)
-      val entityType = WikiTablesTypeDeclaration.CELL_TYPE
+      val entityString = linking.getSecond.toString
```
Comment (jayantk) [inline src/main/scala/org/allenai/wikitables/WikiTablesSemanticParserCli.scala]: is .toString() really the right way to get the entity name?

File: src/main/scala/org/allenai/wikitables/WikiTablesSemanticParserCli.scala [hunk 7]
```diff
@@ -170,19 +190,31 @@ class WikiTablesSemanticParserCli extends AbstractCli() {
...
+      val entityExpr = Expression2.constant(entityString)
+      val entityType = typeDeclaration.getType(entityString)
       val template = ConstantTemplate(entityType, entityExpr)
       // TODO: Passing names as null. Not sure what happens.
       val entity = Entity(entityExpr, entityType, template, null)
       // Note: Passing a constant score of 0.1 for all matches
-      builder += ((Span(linking.getFirst().getFirst(), linking.getFirst().getSecond()), entity, tokenIds, 0.1))
+      if (linking.getFirst() == null) {
+        val entityParts = entityString.split('.')
```
Comment (jayantk) [inline src/main/scala/org/allenai/wikitables/WikiTablesSemanticParserCli.scala]: Add a comment explaining what these string operations are doing.

**Comment by jayantk** [comment]
Two more things:

1) Is this still running sempre to generate the training data set? It seems like that's going to be way to slow and it'll be easier to just use the data set that Ice gave us.
2) Can you give us instructions on how to run the experiment that you're running?

File: src/main/scala/org/allenai/wikitables/WikiTablesSemanticParserCli.scala (hunk 5 above)
Comment (pdasigi) [inline src/main/scala/org/allenai/wikitables/WikiTablesSemanticParserCli.scala]: So names is a list of list of ints. I'm assuming each inner list is a sequence of token ids that correspond to the entity?

**Comment by jayantk** [comment]
LGTM

**Pull request merged**